src/Compilers/CSharp/Test/EndToEnd/EndToEndTests.cs

```diff
@@ -514,5 +514,125 @@ string makeExpectedOutput()
+            */
+            var builder1 = new StringBuilder();
+            var interfacesNumber = Math.Max(overloads1Number, overloads2Number);
+            for (int i = 0; i < interfacesNumber; i++)
+            {
+                builder1.AppendLine($$"""interface I{{i}} { }""");
+            }
+
+            /*
+                void M1(System.Action<I0> a) { }
+                // ...
+                void M1(System.Action<I9> a) { }
+            */
+            var builder2 = new StringBuilder();
+            for (int i = 0; i < overloads1Number; i++)
+            {
+                builder2.AppendLine($$"""void M1(System.Action<I{{i}}> a) { }""");
+            }
+
+            /*
+                void M2(I0 x, System.Func<string, I0> f) { }
+                // ...
+                void M2(I9 x, System.Func<string, I9> f) { }
+            */
+            for (int i = 0; i < overloads2Number; i++)
+            {
+                builder2.AppendLine($$"""void M2(I{{i}} x, System.Func<string, I{{i}}> f) { }""");
+            }
+
+            var builder3 = new StringBuilder();
+            // Local functions should be similarly fast as lambdas.
+            if (localFunctions)
+            {
+                /*
+                    M2(x, L0);
+                    static I0 L0(string arg) {
+                        arg = arg + "0";
+                        // ...
+                        arg = arg + "9";
+                        return default;
+                    }
+                    // ...
+                    M2(x, L9);
+                    static I0 L9(string arg) { ... }
+                */
+                for (int i = 0; i < lambdasNumber; i++)
+                {
+                    builder3.AppendLine($"M2(x, L{i});");
+                    builder3.AppendLine($$"""static I0 L{{i}}(string arg) {""");
+                    for (int j = 0; j < statementsNumber; j++)
+                    {
+                        builder3.AppendLine($"""arg = arg + "{j}";""");
+                    }
+                    builder3.AppendLine("return default;");
+                    builder3.AppendLine("}");
+                }
+            }
+            else
+            {
+                /*
+                    M2(x, static I0 (string arg) => {
+                        arg = arg + "0";
+                        // ...
+                        arg = arg + "9";
+                        return default;
+                    });
+                    // ...
+                    M2(x, static I0 (string arg) => {
+                        // ...
+                        return default;
+                    });
+                */
+                for (int i = 0; i < lambdasNumber; i++)
+                {
+                    builder3.AppendLine("M2(x, static I0 (string arg) => {");
+                    for (int j = 0; j < statementsNumber; j++)
+                    {
+                        builder3.AppendLine($"""arg = arg + "{j}";""");
+                    }
```

Comment:
What is the effect of `lamdasNumber` or `statementsNumber`? Are these just constant multiples for the total execution time with or without the fix? If so, would it make sense to simply use 1 lambda and 1 statement?

Reply:
With this change of the test (time doesn't change visibly anymore):

| | lambda bindings | time |
|--------|--------|--------|
| local functions | 20 | ~1s |
| lambdas | 211 → 40 | ~0.1s |